File: src/Functional/Haskell/Bridge.hs
module Main where

data Device = Device { turnOn :: String, mute :: String }

tv, radio :: Device
tv = Device "TV:on" "TV:muted"
radio = Device "Radio:on" "Radio:muted"

type Remote = Device -> String

basicRemote, muteRemote :: Remote
basicRemote = turnOn
muteRemote = mute

main :: IO ()
main = do
  putStrLn $ "basic-tv=" ++ basicRemote tv
  putStrLn $ "basic-radio=" ++ basicRemote radio
  putStrLn $ "mute-tv=" ++ muteRemote tv
  putStrLn $ "mute-radio=" ++ muteRemote radio

File: src/Functional/Scala/Bridge.scala
trait Device {
  def turnOn: String
  def mute: String
}

final case class NamedDevice(name: String) extends Device {
  override def turnOn: String = s"$name:on"
  override def mute: String = s"$name:muted"
}

sealed trait Remote {
  def execute(device: Device): String
}

case object BasicRemote extends Remote {
  override def execute(device: Device): String = device.turnOn
}

case object MuteRemote extends Remote {
  override def execute(device: Device): String = device.mute
}

object Bridge {
  def main(args: Array[String]): Unit = {
    val tv = NamedDevice("TV")
    val radio = NamedDevice("Radio")
    println(s"basic-tv=${BasicRemote.execute(tv)}")
    println(s"basic-radio=${BasicRemote.execute(radio)}")
    println(s"mute-tv=${MuteRemote.execute(tv)}")
    println(s"mute-radio=${MuteRemote.execute(radio)}")
  }
}
